README.md
# Face Recognition
Face Recognition Based on Facenet

Built using [Facenet](https://github.com/davidsandberg/facenet)'s state-of-the-art face recognition
built with deep learning. The model has an accuracy of 99.2% on the
[Labeled Faces in the Wild](http://vis-www.cs.umass.edu/lfw/) benchmark.

## Features
* Out of Box Working Face Recognition 
* Choose Any Pre-Trained Model from Facenet
* For training just provide the proper folder structure
* Faster than other available solutions

### Prerequisites

* You need Python(2.6 to 3.5) installed
* X-based System supported _(does work on Windows but not tested)_

### Installing

```python
pip install facenet_recognition
```


#### Setup 

__Create setup as follows:__

![input_dir](https://user-images.githubusercontent.com/15515106/38088234-b267317e-3378-11e8-9e82-e4615d7568d3.png)


1. Create input directory eg: input_images
![screen shot 2018-03-29 at 5 42 47 pm](https://user-images.githubusercontent.com/15515106/38088332-11e51f4e-3379-11e8-9db4-90e57b0f6b7f.png)

__We require at least an image per person__

2. Create aligned images directory eg: aligned_images
_Create this directory we will store aligned images here_
3. Create pre-trained model directory eg: pretrained_facenet_model
_Download Pre-Trained model from [Facenet](https://github.com/davidsandberg/facenet)_ and keep it in the pre_model directory
4. Create my trained classifier directory eg: my_classifier
_In this directory we will save our trained model_


## Let's Begin

#### For Facial Recognition we need to align images as follows:
```python
import facenet_recognition
facenet_recognition.align_input('input_images','aligned_images')
```
_Above command will create our input images into aligned format and save it in given aligned images folder_

#### Train & Test Classifier on Images

After we have aligned images now we can train our classifier.

```python
pre_model='./pretrained_facenet_model/20170511-185253.pb' #locaiton of pret-trained model from Facenet
my_class ='./my_classifier/my_classifier.pkl' #location where we want to save
test_classifier_type = 'svm' #type of model either svm or nn
weight= './my_classifier/model_small.yaml' #local stored weights

facenet_recognition.test_train_classifier(aligned_images,pre_model,my_class,weight,test_classifier_type,nrof_train_images_per_class=30, seed=102)

```
_Mininum Required Image per person_: _1_
_Number of Images for Training per Person_: _30 (configurable)_

#### Train Classifer on Images(only Training)

This API is used to Train our Classifier on Aligned Images

```python
pre_model='./pretrained_facenet_model/20170511-185253.pb' #locaiton of pret-trained model from Facenet
my_class ='./my_classifier/my_classifier.pkl' #location where we want to save
test_classifier_type = 'nn' #type of model either svm or nn
weight= './my_classifier/model_small.yaml' #local stored weights

facenet_recognition.create_classifier(aligned_images,pre_model,my_class,weight,test_classifier_type)

```
_Mininum Required Image per person_: _1_
_Number of Images for Training per Person_: _30 (fixed)_

#### Test Classifer on Images
This API is used to test our Trained Classifer
```python
pre_model='./pretrained_facenet_model/20170511-185253.pb' #locaiton of pret-trained model from Facenet
my_class ='./my_classifier/my_classifier.pkl' #location where we want to save
test_classifier_type = 'nn' #type of model either svm or nn
weight= './my_classifier/model_small.yaml' #local stored weights

facenet_recognition.test_classifier(aligned_images,pre_model,my_class,weight,test_classifier_type)

```
_Mininum Required Image per person_: _1_

## Authors

* **Ishwar Sawale** -- [Visit Portfolio ](http://ishwarsawale.com)

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details

## Acknowledgments

* Big Thanks to David Sandberg for Facent
* Inspired by Dlib based library face_recognition

HISTORY.rst
History
=======

0.1.4 (2018-28-03)
------------------

* First beta release.


facenet_recognition/auto_train_test.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import tensorflow as tf
import numpy as np
import argparse
# import facenet
from facenet_recognition import facenet
import os
import sys
import math
import pickle
from sklearn.svm import SVC
from sklearn.cluster import KMeans
from numpy import array
from keras.models import Sequential
from keras.layers import Convolution2D
from keras.layers import MaxPooling2D
from keras.layers import Flatten
from keras.layers.core import Dense, Dropout, Activation
from keras.models import load_model
from keras.models import model_from_yaml
from sklearn.model_selection import StratifiedKFold
import keras
# import align_images as align
from facenet_recognition import align_images as align
from datetime import datetime
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

def train_test(mode,classifier_filename, model_type,min_nrof_images_per_class,nrof_train_images_per_class,
               batch_size,image_size,use_split_dataset,model_dir,seed,data_dir=None, classifier_weight=None):
    # seed=102
    # min_nrof_images_per_class=1
    # nrof_train_images_per_class=30
    # batch_size=1000
    # image_size=160
    # data_dir='out_dir' 
    # use_split_dataset=False
    # model_dir = './pre_model/20170511-185253.pb'
    classe_total = len([name for name in os.listdir(data_dir)])

    with tf.Graph().as_default():

        with tf.Session() as sess:

            np.random.seed(seed=seed)

            if use_split_dataset:
                dataset_tmp = facenet.get_dataset(data_dir)
                train_set, test_set = split_dataset(dataset_tmp, min_nrof_images_per_class, nrof_train_images_per_class)
                if (mode=='TRAIN'):
                    dataset = train_set
                elif (mode=='TEST'):
                    dataset = test_set
            else:
                dataset = facenet.get_dataset(data_dir)
            paths, labels = facenet.get_image_paths_and_labels(dataset)
            print('Number of classes: %d' % len(dataset))
            print('Number of images: %d' % len(paths))

            # Load the model
            print('Loading feature extraction model')
            startTimePre = datetime.now()
            facenet.load_model(model_dir)

            # Get input and output tensors
            images_placeholder = tf.get_default_graph().get_tensor_by_name("input:0")
            embeddings = tf.get_default_graph().get_tensor_by_name("embeddings:0")
            phase_train_placeholder = tf.get_default_graph().get_tensor_by_name("phase_train:0")
            embedding_size = embeddings.get_shape()[1]

            # Run forward pass to calculate embeddings
            print('Calculating features for images')
            nrof_images = len(paths)
            nrof_batches_per_epoch = int(math.ceil(1.0*nrof_images / batch_size))
            emb_array = np.zeros((nrof_images, embedding_size))
            for i in range(nrof_batches_per_epoch):
                start_index = i*batch_size
                end_index = min((i+1)*batch_size, nrof_images)
                paths_batch = paths[start_index:end_index]
                images = facenet.load_data(paths_batch, False, False, image_size)
                feed_dict = { images_placeholder:images, phase_train_placeholder:False }
                emb_array[start_index:end_index,:] = sess.run(embeddings, feed_dict=feed_dict)

            endTimePre = datetime.now() - startTimePre    
            classifier_filename_exp = os.path.expanduser(classifier_filename)

            if (mode=='TRAIN'):
                # Train classifier
                startTime = datetime.now()
                if(model_type=='SVM'):
                    print('Training classifier')
                    model = SVC(kernel='linear', probability=True)
                    model.fit(emb_array, labels)

                    # Create a list of class names
                    class_names = [ cls.name.replace('_', ' ') for cls in dataset]

                    # Saving classifier model
                    with open(classifier_filename_exp, 'wb') as outfile:
                        pickle.dump((model, class_names), outfile)
                    print('Saved classifier model to file "%s"' % classifier_filename_exp)
                if(model_type=='NN'):
                    label_array = array(labels)
                    one_hot_labels = keras.utils.to_categorical(label_array, num_classes=classe_total)
                    print('Training classifier')
                    model = Sequential()
                    model.add(Dense(32, activation='relu', input_dim=128))
                    model.add(Dense(16, activation='relu'))
                    model.add(Dropout(0.5))
                    model.add(Dense(classe_total, activation='softmax'))
                    model.compile(optimizer='rmsprop',
                                loss='categorical_crossentropy',
                                metrics=['accuracy'])
                    model.fit(emb_array, one_hot_labels, epochs=100, batch_size=32)
                    print('trainging finished for NN Facenet Model')
                    scores = model.evaluate(emb_array, one_hot_labels, verbose=0)
                    print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))
                    # serialize model to YAML
                    model_yaml = model.to_yaml()
                    with open(classifier_weight, 'wb') as yaml_file:
                        yaml_file.write(model_yaml)
                    model.save(classifier_filename)
                    print("Saved model to disk")
                endTime = datetime.now() - startTime    
                print('It took', endTime, 'seconds after embedding to train model')
                print('It took', endTimePre, 'seconds time required to calculate embedding for')
            elif (mode=='TEST'):
                # Classify images
                if(model_type == "SVM"):
                    print('Testing classifier')
                    with open(classifier_filename_exp, 'rb') as infile:
                        (model, class_names) = pickle.load(infile)
                    print('Loaded classifier model from file "%s"' % classifier_filename_exp)
                    predictions = model.predict_proba(emb_array)
                    best_class_indices = np.argmax(predictions, axis=1)
                    best_class_probabilities = predictions[np.arange(len(best_class_indices)), best_class_indices]
                    for i in range(len(best_class_indices)):
                        print('%4d  %s: %.3f' % (i, class_names[best_class_indices[i]], best_class_probabilities[i]))
                    accuracy = np.mean(np.equal(best_class_indices, labels))
                    print('Accuracy: %.3f' % accuracy)
                if(model_type=="NN"):
                    label_array = array(labels)
                    one_hot_labels = keras.utils.to_categorical(label_array, num_classes=classe_total)
                    yaml_file = open('./my_class/model_small.yaml', 'r')
                    loaded_model_yaml = yaml_file.read()
                    yaml_file.close()
                    loaded_model = model_from_yaml(loaded_model_yaml)
                    # load weights into new model
                    loaded_model.load_weights(classifier_filename)
                    loaded_model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])
                    print("Loaded model from disk")
                    score = loaded_model.evaluate(emb_array, one_hot_labels, verbose=0)
                    print("%s: %.2f%%" % (loaded_model.metrics_names[1], score[1]*100))
                

def split_dataset(dataset, min_nrof_images_per_class, nrof_train_images_per_class):
    train_set = []
    test_set = []
    for cls in dataset:
        paths = cls.image_paths
        # Remove classes with less than min_nrof_images_per_class
        if len(paths)>=min_nrof_images_per_class:
            np.random.shuffle(paths)
            train_set.append(facenet.ImageClass(cls.name, paths[:nrof_train_images_per_class]))
            test_set.append(facenet.ImageClass(cls.name, paths[nrof_train_images_per_class:]))
    return train_set, test_set



# def nn(option):
#     if (option == 'TRAIN'):
#         train_test('TRAIN', './my_class/my_model_small.h5', 'NN')
#     else:
#         train_test('TEST', './my_class/my_model_small.h5', 'NN')

# def svm(option):
#     if(option=='TRAIN'):
#         train_test('TRAIN', './my_class/my_classifier.pkl', 'SVM')
#     if(option=='TEST'):
#         train_test('TEST', './my_class/my_classifier.pkl', 'SVM')
#     if(option=='both'):
#         train_test('TRAIN', './my_class/my_classifier.pkl', 'SVM')
#         train_test('TEST', './my_class/my_classifier.pkl', 'SVM')
        

# if __name__ == '__main__':
#     startTime = datetime.now()
#     # SVM('TRAIN')
#     NN('TRAIN')
#     # align.align_data('./input_dir', './out_dir')
#     print(datetime.now() - startTime) 
#     # NN('TEST')

facenet_recognition/api.py
# -*- coding: utf-8 -*-
from facenet_recognition import align_images as align
from facenet_recognition import auto_train_test as classifier
import os, os.path

def pre_check(directory):
    if not os.path.exists(directory):
        raise(directory, 'does not exist')

def align_input(input_dir=None, out_dir=None):
    pre_check(input_dir)
    pre_check(out_dir)
    align.align_data(input_dir, out_dir)

def create_classifier(input_dir=None, model_dir=None,classifier_dir=None, classifier_weight=None, model_type=None):
    pre_check(input_dir)
    if model_type.upper() == 'SVM':
        classifier.train_test(
            mode='TRAIN',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=30,
            batch_size=1000,image_size=160,
            use_split_dataset=False,
            model_dir = model_dir,
            seed=102, data_dir=input_dir)
    elif model_type.upper() == 'NN' and classifier_weight:
        classifier.train_test(
            mode='TRAIN',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=30,
            batch_size=1000,image_size=160,
            use_split_dataset=False,
            model_dir = model_dir,
            seed=102, data_dir=input_dir,
            classifier_weight=classifier_weight)
    else:
        print('only svm and nn classifier are supported')

def test_train_classifier(input_dir=None, model_dir=None,classifier_dir=None, classifier_weight=None, model_type=None,
    nrof_train_images_per_class=30, seed=102):
    pre_check(input_dir)
    print(seed)
    if model_type.upper() == 'SVM':
        classifier.train_test(
            mode='TRAIN',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=nrof_train_images_per_class,
            batch_size=1000,image_size=160,
            use_split_dataset=True,
            model_dir = model_dir,
            seed=seed, data_dir=input_dir)
        classifier.train_test(
            mode='TEST',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=nrof_train_images_per_class,
            batch_size=1000,image_size=160,
            use_split_dataset=True,
            model_dir = model_dir,
            seed=seed, data_dir=input_dir)
    elif model_type.upper() == 'NN' and classifier_weight:
        classifier.train_test(
            mode='TRAIN',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=nrof_train_images_per_class,
            batch_size=1000,image_size=160,
            use_split_dataset=True,
            model_dir = model_dir,
            seed=seed, data_dir=input_dir,
            classifier_weight=classifier_weight)
        classifier.train_test(
            mode='TEST',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=nrof_train_images_per_class,
            batch_size=1000,image_size=160,
            use_split_dataset=True,
            model_dir = model_dir,
            seed=seed, data_dir=input_dir,
            classifier_weight=classifier_weight)
    else:
        print('only svm and nn classifier are supported')

def test_classifier(input_dir=None, model_dir=None,classifier_dir=None, classifier_weight=None, model_type=None):
    pre_check(input_dir)
    if model_type.upper() == 'SVM':
        classifier.train_test(
            mode='TEST',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=30,
            batch_size=1000,image_size=160,
            use_split_dataset=False,
            model_dir = model_dir,
            seed=102, data_dir=input_dir)
    elif model_type.upper() == 'NN' and classifier_weight:
        classifier.train_test(
            mode='TEST',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=30,
            batch_size=1000,image_size=160,
            use_split_dataset=False,
            model_dir = model_dir,
            seed=102, data_dir=input_dir,
            classifier_weight=classifier_weight)
    else:
        print('only svm and nn classifier are supported')

facenet_recognition/face.py
# -*- coding: utf-8 -*-
import align_images as align
import auto_train_test as classifier
import os, os.path

def pre_check(directory):
    if not os.path.exists(directory):
        raise(directory, 'does not exist')

def align_input(input_dir=None, out_dir=None):
    pre_check(input_dir)
    pre_check(out_dir)
    align.align_data(input_dir, out_dir)

def create_classifier(input_dir=None, model_dir=None,classifier_dir=None, classifier_weight=None, model_type=None):
    pre_check(input_dir)
    if model_type.upper() == 'SVM':
        classifier.train_test(
            mode='TRAIN',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=30,
            batch_size=1000,image_size=160,
            use_split_dataset=False,
            model_dir = model_dir,
            seed=102, data_dir=input_dir)
    elif model_type.upper() == 'NN' and classifier_weight:
        classifier.train_test(
            mode='TRAIN',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=30,
            batch_size=1000,image_size=160,
            use_split_dataset=False,
            model_dir = model_dir,
            seed=102, data_dir=input_dir,
            classifier_weight=classifier_weight)
    else:
        print('only svm and nn classifier are supported')

def test_train_classifier(input_dir=None, model_dir=None,classifier_dir=None, classifier_weight=None, model_type=None,
    nrof_train_images_per_class=30, seed=102):
    pre_check(input_dir)
    print(seed)
    if model_type.upper() == 'SVM':
        classifier.train_test(
            mode='TRAIN',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=nrof_train_images_per_class,
            batch_size=1000,image_size=160,
            use_split_dataset=True,
            model_dir = model_dir,
            seed=seed, data_dir=input_dir)
        classifier.train_test(
            mode='TEST',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=nrof_train_images_per_class,
            batch_size=1000,image_size=160,
            use_split_dataset=True,
            model_dir = model_dir,
            seed=seed, data_dir=input_dir)
    elif model_type.upper() == 'NN' and classifier_weight:
        classifier.train_test(
            mode='TRAIN',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=nrof_train_images_per_class,
            batch_size=1000,image_size=160,
            use_split_dataset=True,
            model_dir = model_dir,
            seed=seed, data_dir=input_dir,
            classifier_weight=classifier_weight)
        classifier.train_test(
            mode='TEST',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=nrof_train_images_per_class,
            batch_size=1000,image_size=160,
            use_split_dataset=True,
            model_dir = model_dir,
            seed=seed, data_dir=input_dir,
            classifier_weight=classifier_weight)
    else:
        print('only svm and nn classifier are supported')

def test_classifier(input_dir=None, model_dir=None,classifier_dir=None, classifier_weight=None, model_type=None):
    pre_check(input_dir)
    if model_type.upper() == 'SVM':
        classifier.train_test(
            mode='TEST',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=30,
            batch_size=1000,image_size=160,
            use_split_dataset=False,
            model_dir = model_dir,
            seed=102, data_dir=input_dir)
    elif model_type.upper() == 'NN' and classifier_weight:
        classifier.train_test(
            mode='TEST',classifier_filename=classifier_dir, 
            model_type=model_type.upper(),min_nrof_images_per_class=1,
            nrof_train_images_per_class=30,
            batch_size=1000,image_size=160,
            use_split_dataset=False,
            model_dir = model_dir,
            seed=102, data_dir=input_dir,
            classifier_weight=classifier_weight)
    else:
        print('only svm and nn classifier are supported')

facenet_recognition/align_images.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from scipy import misc
import sys
import os
import argparse
import tensorflow as tf
import numpy as np
# import facenet
from facenet_recognition import facenet
# import detect_face
from facenet_recognition import detect_face
import random
from time import sleep
from datetime import datetime
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'


def align_data(input_dir='./input_dir/', output_dir='./out_dir'):
    output_dir_path = output_dir
    output_dir = os.path.expanduser(output_dir_path)
    if not os.path.exists(output_dir):
            os.makedirs(output_dir)
    datadir = input_dir
    dataset = facenet.get_dataset(datadir)
    startTime = datetime.now()
    print('Creating networks and loading parameters')
    with tf.Graph().as_default():
        gpu_options = tf.GPUOptions(per_process_gpu_memory_fraction=0.5)
        sess = tf.Session(config=tf.ConfigProto(gpu_options=gpu_options, log_device_placement=False))
        with sess.as_default():
            pnet, rnet, onet = detect_face.create_mtcnn(sess)

    minsize = 20  # minimum size of face
    threshold = [0.6, 0.7, 0.7]  # three steps's threshold
    factor = 0.709  # scale factor
    margin = 44
    image_size = 182

    # Add a random key to the filename to allow alignment using multiple processes
    random_key = np.random.randint(0, high=99999)
    bounding_boxes_filename = os.path.join(output_dir, 'bounding_boxes_%05d.txt' % random_key)
    print('Goodluck')

    with open(bounding_boxes_filename, "w") as text_file:
        nrof_images_total = 0
        nrof_successfully_aligned = 0
        for cls in dataset:
            output_class_dir = os.path.join(output_dir, cls.name)
            if not os.path.exists(output_class_dir):
                os.makedirs(output_class_dir)
            for image_path in cls.image_paths:
                nrof_images_total += 1
                filename = os.path.splitext(os.path.split(image_path)[1])[0]
                output_filename = os.path.join(output_class_dir, filename + '.png')
                print(image_path)
                if not os.path.exists(output_filename):
                    try:
                        img = misc.imread(image_path)
                        print('read data dimension: ', img.ndim)
                    except (IOError, ValueError, IndexError) as e:
                        errorMessage = '{}: {}'.format(image_path, e)
                        print(errorMessage)
                    else:
                        if img.ndim < 2:
                            print('Unable to align "%s"' % image_path)
                            # text_file.write('%s\n' % (output_filename))
                            continue
                        if img.ndim == 2:
                            img = facenet.to_rgb(img)
                            print('to_rgb data dimension: ', img.ndim)
                        img = img[:, :, 0:3]
                        print('after data dimension: ', img.ndim)

                        bounding_boxes, _ = detect_face.detect_face(img, minsize, pnet, rnet, onet, threshold, factor)
                        nrof_faces = bounding_boxes.shape[0]
                        print('detected_face: %d' % nrof_faces)
                        if nrof_faces > 0:
                            det = bounding_boxes[:, 0:4]
                            img_size = np.asarray(img.shape)[0:2]
                            if nrof_faces > 1:
                                bounding_box_size = (det[:, 2] - det[:, 0]) * (det[:, 3] - det[:, 1])
                                img_center = img_size / 2
                                offsets = np.vstack([(det[:, 0] + det[:, 2]) / 2 - img_center[1],
                                                     (det[:, 1] + det[:, 3]) / 2 - img_center[0]])
                                offset_dist_squared = np.sum(np.power(offsets, 2.0), 0)
                                index = np.argmax(bounding_box_size - offset_dist_squared * 2.0)  # some extra weight on the centering
                                det = det[index, :]
                            det = np.squeeze(det)
                            bb_temp = np.zeros(4, dtype=np.int32)

                            bb_temp[0] = det[0]
                            bb_temp[1] = det[1]
                            bb_temp[2] = det[2]
                            bb_temp[3] = det[3]
                            try:
                                cropped_temp = img[bb_temp[1]:bb_temp[3], bb_temp[0]:bb_temp[2], :]
                                scaled_temp = misc.imresize(cropped_temp, (image_size, image_size), interp='bilinear')
                                nrof_successfully_aligned += 1
                                misc.imsave(output_filename, scaled_temp)
                            # text_file.write('%s %d %d %d %d\n' % (output_filename, bb_temp[0], bb_temp[1], bb_temp[2], bb_temp[3]))
                            except Exception as e:
                                os.remove(image_path)
                                
                           
                        else:
                            print('Unable to align "%s"' % image_path)
                            os.remove(image_path)
                            print('removed file %s' % image_path)
                            # text_file.write('%s\n' % (output_filename))
    print('Total number of images: %d' % nrof_images_total)
    print('Number of successfully aligned images: %d' % nrof_successfully_aligned)
    os.remove(bounding_boxes_filename)
    print(datetime.now() - startTime) 
    
# align_data('/Users/ishwarsawale/working_demos/face_rec_demo/input_dir',
# '/Users/ishwarsawale/working_demos/face_rec_demo/out_dir')

facenet_recognition/nn_webcam_demo.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import tensorflow as tf
from scipy import misc
import cv2
import matplotlib.pyplot as plt
import numpy as np
import argparse
import facenet
import detect_face
import os
from os.path import join as pjoin
import sys
import time
import copy
import math
import pickle
from sklearn.svm import SVC
from sklearn.externals import joblib

print('Creating networks and loading parameters')
with tf.Graph().as_default():
    gpu_options = tf.GPUOptions(per_process_gpu_memory_fraction=0.6)
    sess = tf.Session(config=tf.ConfigProto(gpu_options=gpu_options, log_device_placement=False))
    with sess.as_default():
        pnet, rnet, onet = detect_face.create_mtcnn(sess, './d_npy')

        minsize = 20  # minimum size of face
        threshold = [0.6, 0.7, 0.7]  # three steps's threshold
        factor = 0.709  # scale factor
        margin = 44
        frame_interval = 3
        batch_size = 1000
        image_size = 182
        input_image_size = 160
        
        HumanNames = os.listdir("./input_dir")
        HumanNames.sort()

        print('Loading feature extraction model')
        modeldir = './pre_model/20170511-185253.pb'
        facenet.load_model(modeldir)

        images_placeholder = tf.get_default_graph().get_tensor_by_name("input:0")
        embeddings = tf.get_default_graph().get_tensor_by_name("embeddings:0")
        phase_train_placeholder = tf.get_default_graph().get_tensor_by_name("phase_train:0")
        embedding_size = embeddings.get_shape()[1]

        classifier_filename = './my_class/my_classifier.pkl'
        classifier_filename_exp = os.path.expanduser(classifier_filename)
        with open(classifier_filename_exp, 'rb') as infile:
            (model, class_names) = pickle.load(infile)
            print('load classifier file-> %s' % classifier_filename_exp)

        video_capture = cv2.VideoCapture(0)
        c = 0
        print('Start Recognition!')
        prevTime = 0
        while True:
            ret, frame = video_capture.read()

            frame = cv2.resize(frame, (0,0), fx=0.5, fy=0.5)    #resize frame (optional)

            curTime = time.time()+1    # calc fps
            timeF = frame_interval

            if (c % timeF == 0):
                find_results = []

                if frame.ndim == 2:
                    frame = facenet.to_rgb(frame)
                frame = frame[:, :, 0:3]
                bounding_boxes, _ = detect_face.detect_face(frame, minsize, pnet, rnet, onet, threshold, factor)
                nrof_faces = bounding_boxes.shape[0]
                print('Detected_FaceNum: %d' % nrof_faces)

                if nrof_faces > 0:
                    det = bounding_boxes[:, 0:4]
                    img_size = np.asarray(frame.shape)[0:2]

                    cropped = []
                    scaled = []
                    scaled_reshape = []
                    bb = np.zeros((nrof_faces,4), dtype=np.int32)

                    for i in range(nrof_faces):
                        emb_array = np.zeros((1, embedding_size))

                        bb[i][0] = det[i][0]
                        bb[i][1] = det[i][1]
                        bb[i][2] = det[i][2]
                        bb[i][3] = det[i][3]

                        # inner exception
                        if bb[i][0] <= 0 or bb[i][1] <= 0 or bb[i][2] >= len(frame[0]) or bb[i][3] >= len(frame):
                            print('face is inner of range!')
                            continue

                        cropped.append(frame[bb[i][1]:bb[i][3], bb[i][0]:bb[i][2], :])
                        cropped[i] = facenet.flip(cropped[i], False)
                        scaled.append(misc.imresize(cropped[i], (image_size, image_size), interp='bilinear'))
                        scaled[i] = cv2.resize(scaled[i], (input_image_size,input_image_size),
                                               interpolation=cv2.INTER_CUBIC)
                        scaled[i] = facenet.prewhiten(scaled[i])
                        scaled_reshape.append(scaled[i].reshape(-1,input_image_size,input_image_size,3))
                        feed_dict = {images_placeholder: scaled_reshape[i], phase_train_placeholder: False}
                        emb_array[0, :] = sess.run(embeddings, feed_dict=feed_dict)
                        predictions = model.predict_proba(emb_array)
                        print(predictions)
                        best_class_indices = np.argmax(predictions, axis=1)
                        print(best_class_indices)
                        best_class_probabilities = predictions[np.arange(len(best_class_indices)), best_class_indices]
                        print(best_class_probabilities)
                        cv2.rectangle(frame, (bb[i][0], bb[i][1]), (bb[i][2], bb[i][3]), (0, 255, 0), 2)    #boxing face

                        #plot result idx under box
                        text_x = bb[i][0]
                        text_y = bb[i][3] + 20
                        print('result: ', best_class_indices[0])
                        print(best_class_indices)
                        print(HumanNames)
                        for H_i in HumanNames:
                            print(H_i)
                            if HumanNames[best_class_indices[0]] == H_i:
                                result_names = HumanNames[best_class_indices[0]]
                                cv2.putText(frame, result_names, (text_x, text_y), cv2.FONT_HERSHEY_COMPLEX_SMALL,
                                            1, (0, 0, 255), thickness=1, lineType=2)
                else:
                    print('Unable to align')

            sec = curTime - prevTime
            prevTime = curTime
            fps = 1 / (sec)
            str = 'FPS: %2.3f' % fps
            text_fps_x = len(frame[0]) - 150
            text_fps_y = 20
            cv2.putText(frame, str, (text_fps_x, text_fps_y),
                        cv2.FONT_HERSHEY_COMPLEX_SMALL, 1, (0, 0, 0), thickness=1, lineType=2)
            # c+=1
            cv2.imshow('Video', frame)

            if cv2.waitKey(1) & 0xFF == ord('q'):
                break

        video_capture.release()
        # #video writer
        cv2.destroyAllWindows()